cornellh4i/okb-hope
# Dynamic routes implementation

## Overview


- Set up dynamic routes so that we can go to something like /path/to/website/{user}/dashboard
- Fetch messages based on the user logged in
- Display dashboard based on the user logged in
- Linked Discover Professionals page to the professional profile
- Linked dashboard page to edit profile page
- Link edit profile to backend based on user logged in
- Display Appointments based on the user logged in, along with responses pulled from Patient profile information

## Replication


```
cd frontend
yarn dev
```
- Log in as user
- Check messages
- Check dashboard's "Saved Psychiatrists" section
- Check link between Discover page and Prof Profile works
- Edit profile and check that edits are updated to Firebase
- Check dashboard's "Appointments" section, check "Appointment Details"
## Changes Made


`Week 1`
- Set up dynamic routes for all the user-dependent pages (`discover.tsx`, `dashboard.tsx`, `edit_profile.tsx`, `index.tsx`, `messages.tsx`, `prof_profile.tsx`) so that we can go to something like /path/to/website/{user}/dashboard

`Week 2`
- Fetch messages based on user logged in (modified `MessageList.tsx`)
- Added index to Firebase's Cloud Firestore in order for messages based on user logged in AND placed in order of creation 
- Updated `IUser` in `schema.ts`
- Fetch savedPsychiatrists and display on dashboard based on user logged in (modified `PsychiatristCardsListing.tsx`)
- Added temporary function to add fields to user collection until `users` collection issue can be resolved
- Fixed schema-related issues in `schema.ts`, `fetchData.ts`, `ProfProfile.tsx`, `PsychiatristList.tsx`, `[userId]/discover.tsx`
- Linked Discover Professionals page to the professional profile
- Linked dashboard page to edit profile page
- Polished frontend of edit profile page / ability to select options 

`Week 3`
- Fetch Patient profile information from Firebase into Edit Profile and update Firebase information accordingly when changes are made
- Fetch Appointment information for user logged in - pull data from Patient profile to populate Appointment questions on "Appointment Details" popup, pull data from `psychiatrists` collection by querying on `profId` field in `appointments` collection - edited `AppointmentList.tsx`, `AppointmentCard.tsx`, `AppointmentQuestion.tsx`
- Added `fetchApptDetails` function to `fetchData.ts`
- Changed `apptId` to `appointId` in `schema.ts` to keep consistent with Firebase field
## Test Coverage


- Log into a user account and check that the route is correct (includes the user id) when you go to different pages. 
- Check to see if you can only see messages sent from yourself. 
- Used different accounts to test if you can only see your savedPsychiatrists (check savedPsychiatrist field in firebase)
- Made edits to all fields of various patients' profiles and checked that Firebase is updated accordingly
- Tested to see if the Discover Professionals page correctly links between itself and Professional Profiles while maintaining the dynamic user route
- Added data in `appointments` collection in firebase to see if it renders on dashboard
## Next Steps 


Week 2:
- Linking the backend (fetch all the information based on the user logged in for messages, dashboard)
- Link Discover Professionals page to the professional profile
- Link dashboard page to edit profile page

Week 3:
- Linking the backend (edit_profile and appointments). For appointments, we have to alter the db to support which professionals are linked to which clients.
Move messaging bubbles to the right side (frontend)?
- Solidify user (patient) schema in relation to edit_profile

- Figure out why time does not display properly sometimes (wrong date, missing a digit)

## Related PRs or Issues


Closes #76 
Closes #82 
## Screenshots


<img width="554" alt="Screenshot 2023-10-24 at 4 11 43 PM" src="https://github.com/cornellh4i/okb-hope/assets/123599926/de48be7f-b163-407b-8ed6-13847460aadc">
<img width="1417" alt="Screen Shot 2023-10-24 at 4 58 38 PM" src="https://github.com/cornellh4i/okb-hope/assets/123703741/ebbdb65f-404d-49d1-a937-13cf2f001942">
<img width="1470" alt="Screenshot 2023-10-31 at 2 26 57 PM" src="https://github.com/cornellh4i/okb-hope/assets/123599926/c1b4328c-4a0c-4c4c-94b4-ef2909cd8bfa">
<img width="1340" alt="Screen Shot 2023-10-31 at 3 29 31 PM" src="https://github.com/cornellh4i/okb-hope/assets/123703741/421253af-7afd-43ad-83ca-15ce8fc442ec">
<img width="1312" alt="Screen Shot 2023-11-06 at 1 16 09 PM" src="https://github.com/cornellh4i/okb-hope/assets/123703741/16cfcc4c-f821-4fe1-8c75-3782afe593ae">
<img width="627" alt="Screen Shot 2023-11-06 at 1 16 30 PM" src="https://github.com/cornellh4i/okb-hope/assets/123703741/abb77b4a-3def-4b32-8c5d-e8401bc986f0">
<img width="1001" alt="Screen Shot 2023-11-06 at 1 17 35 PM" src="https://github.com/cornellh4i/okb-hope/assets/123703741/89338c0f-32de-4124-b0c6-589e9af2e7e6">


<details>

  <summary>Screen Shot Name</summary>

  

</details>
